src/vm/functions/iterables.rs

```diff
@@ -0,0 +1,205 @@
+                let argument = vec![SymbolicExpression::atom_value(x)];
+                apply(&function, &argument, env, context).unwrap()
+            }).collect();
+            Value::list_from(mapped_vec)
+        },
+        Value::Buffer(mut buff) => {
+            let mapped_vec = buff.data.drain(..).map(|x| {
+                let element = Value::buff_from(vec![x]).unwrap();
+                let argument = vec![SymbolicExpression::atom_value(element)];
+                apply(&function, &argument, env, context).unwrap()
+            }).collect();
+            Value::list_from(mapped_vec)
+        },
+        _ => Err(CheckErrors::ExpectedListOrBuffer(TypeSignature::type_of(&iterable)).into())
+    }
+}
+
+pub fn native_append(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(2, args)?;
+
+    let iterable = eval(&args[0], env, context)?;
+    match iterable {
+        Value::List(list) => {
+            let element = eval(&args[1], env, context)?;
+            let mut data_appended = list.data.clone();
+            data_appended.push(element);
+            Value::list_from(data_appended)
+        },
+        _ => Err(CheckErrors::ExpectedListApplication.into())
+    }
+}
+
+pub fn native_concat(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(2, args)?;
+
+    let iterable = eval(&args[0], env, context)?;
+    match iterable {
+        Value::List(mut lhs) => {
+            let mut res = Vec::new();
+            res.append(&mut lhs.data);
+            let mut rhs = eval(&args[1], env, context)?;
+            if let Value::List(ref mut rhs_data) = rhs {
+                res.append(&mut rhs_data.data);
+                Value::list_from(res)
+            } else {
+                Err(CheckErrors::TypeError(
+                    TypeSignature::type_of(&Value::List(lhs)), 
+                    TypeSignature::type_of(&rhs)).into())
+            }
+        },
+        Value::Buffer(mut lhs) => {
+            let mut res = Vec::new();
+            res.append(&mut lhs.data);
+            let mut rhs = eval(&args[1], env, context)?;
+            if let Value::Buffer(ref mut rhs_data) = rhs {
+                res.append(&mut rhs_data.data);
+                Value::buff_from(res)
+            } else {
+                Err(CheckErrors::TypeError(
+                    TypeSignature::BufferType(res.len().try_into().unwrap()),
+                    TypeSignature::type_of(&rhs)).into())
+            }
+        },
+        _ => Err(CheckErrors::ExpectedListOrBuffer(TypeSignature::type_of(&iterable)).into())
+    }
+}
+
+pub fn native_asserts_max_len(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(2, args)?;
+
+    let iterable = eval(&args[0], env, context)?;
+
+    let expected_len = eval(&args[1], env, context)?;
```

Comment:
You should just pull arg[1] directly from the symbolic representation rather than eval'ing it, since it should be a literal.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1139/commits/d8eab4a97ca807bd8a4279635fd12321473d5d52